lib/containers/k8s.pm

```diff
@@ -17,77 +17,112 @@ use testapi;
+
+    # Apply additional k3s installation options
+    # Note: The install script starts a k3s-server by default, unless INSTALL_K3S_SKIP_START is set to true
+    # For more information see https://rancher.com/docs/k3s/latest/en/installation/install-options/#options-for-installation-with-script
+    my %k3s_args = (
+        INSTALL_K3S_SYMLINK => get_var('K3S_SYMLINK'),
+        INSTALL_K3S_BIN_DIR => get_var('K3S_BIN_DIR'),
+        INSTALL_K3S_CHANNEL => get_var('K3S_CHANNEL'),
+        INSTALL_K3S_VERSION => get_var("K3S_VERSION"),
+        INSTALL_K3S_SKIP_START => 'true',
+        K3S_NODE_NAME => 'k3s-node'
+    );
+
+    # github.com/k3s-io/k3s#5946 - The kubectl delete namespace helm-ns-413 command freezes and does nothing
+    my $disables = '--disable=metrics-server';
+    $disables .= ' --disable-helm-controller' unless (get_var('K3S_ENABLE_HELM_CONTROLLER'));
+
+    while (my ($key, $value) = each %k3s_args) {
+        if ($value) {
+            assert_script_run("export $key=$value");
         }
-        assert_script_run("export INSTALL_K3S_SYMLINK=" . get_var('K3S_SYMLINK')) if (get_var('K3S_SYMLINK'));
-        assert_script_run("export INSTALL_K3S_BIN_DIR=" . get_var('K3S_BIN_DIR')) if (get_var('K3S_BIN_DIR'));
-        assert_script_run("export INSTALL_K3S_CHANNEL=" . get_var('K3S_CHANNEL')) if (get_var('K3S_CHANNEL'));
-        # k3s doesn't like long hostnames like the ones being used in publiccloud
-        assert_script_run("export K3S_NODE_NAME=k3s-node") if (is_public_cloud);
-
-        my $disables = '--disable=metrics-server';
-        $disables .= ' --disable-helm-controller' unless (get_var('K3S_ENABLE_HELM_CONTROLLER'));
+        delete $k3s_args{$key};
+    }
 
-        # github.com/k3s-io/k3s#5946 - The kubectl delete namespace helm-ns-413 command freezes and does nothing
-        # Note: The install script starts a k3s-server by default, unless INSTALL_K3S_SKIP_START is set to true
+    if (get_var('K3S_INSTALL_UPSTREAM') || (is_sle || is_sle_micro || is_leap_micro)) {
         script_retry("curl -sfL https://get.k3s.io  -o install_k3s.sh", timeout => 180, delay => 60, retry => 3);
-        assert_script_run("INSTALL_K3S_SKIP_START=true sh install_k3s.sh $disables", timeout => 300);
+        assert_script_run("sh install_k3s.sh $disables", timeout => 300);
         script_run("rm -f install_k3s.sh");
+        setup_and_check_k3s;
+        return;
     }
 
-    if (get_var('REGISTRY')) {
-        script_run("mkdir -p /etc/rancher/k3s");
-        my $registry = registry_url();
-        assert_script_run "curl " . data_url('containers/registries.yaml') . " -o /etc/rancher/k3s/registries.yaml";
-        file_content_replace("/etc/rancher/k3s/registries.yaml", REGISTRY => $registry);
+    # k3s-install script is already packaged for several products
+    my @pkgs = qw(k3s-install);
+    push @pkgs, 'apparmor-parser' if is_sle('>=15-SP1');
+
+    # k3s might be already installed by default
```

Comment:
We can drop this comment or put it at the beginning when we test for `which k3s`.